source/code/include/scxcorelib/scxsingleton.h

```diff
@@ -58,21 +58,39 @@ namespace SCXCoreLib
             If the instance is not yet created, it first creates it.
 
         */
-        static T& Instance();
+        static T& Instance()
+        {
+            if (0 == s_lockHandle.GetData())
+            {
+                throw SCXInternalErrorException(L"Tried to get a singleton instance before static initialization was completed.", SCXSRCLOCATION);
+            }
+
+            SCXThreadLock lock(*s_lockHandle, true);
+
+            if (0 == s_instance.GetData())
+            {
+                s_instance = new T();
+            }
+
+            return *s_instance;
+        }
 
     protected:
         /** Only instantiated by subclassing. */
-        SCXSingleton(int recursiveLock = 1);
+        SCXSingleton(int recursiveLock = 1) {}
 
     private:
         /** Only instantiated by subclassing. */
-        SCXSingleton(const SCXSingleton&);
+        SCXSingleton(const SCXSingleton&) {}
         /** Only copied by subclassing. */
-        SCXSingleton& operator=(const SCXSingleton&);
+        SCXSingleton& operator=(const SCXSingleton&) { return *this; };
 
         static SCXCoreLib::SCXHandle<T> s_instance; //!< Handle to the singleton instance.
         static SCXCoreLib::SCXHandle< SCXThreadLockHandle > s_lockHandle; //!< Used for locking at creation time.
     };
+
+    template<class T> SCXCoreLib::SCXHandle<T> SCXSingleton<T>::s_instance(0);
+    template<class T> SCXCoreLib::SCXHandle<SCXThreadLockHandle> SCXSingleton<T>::s_lockHandle( new SCXThreadLockHandle(ThreadLockHandleGet(1))); // The 1 says allow recursion
 }
```

Comment:
This absolutely definitely won't work. You didn't try PBUILD on this. The Sun compiler will not accept this.